src/celeritas/global/alongstep/AlongStep.hh

```diff
@@ -108,17 +109,83 @@ inline CELER_FUNCTION void along_step(MH&& msc,
         }
     }
 
-    apply_eloss(track, &local.step_limit);
-    CELER_ASSERT(local.step_limit.step > 0);
-    CELER_ASSERT(local.step_limit.action);
+    if (eloss.is_applicable(track))
+    {
+        using Energy = ParticleTrackView::Energy;
+
+        Energy deposited = eloss.calc_eloss(track, local.step_limit.step);
+
+        if (local.step_limit.action != track.boundary_action()
+            && value_as<Energy>(particle.energy()) - value_as<Energy>(deposited)
+                   <= value_as<Energy>(
+                       track.make_physics_view().scalars().eloss_calc_limit))
+        {
+            // The energy after slowing down is less than the hard tracking
+            // cutoff, so stop the particle rather than leaving it a tiny
+            // amount of energy.
+            // To avoid stopping particles on boundaries (which is an
+            // artificial bias), we avoid this for geometry-limited steps.
+            deposited = particle.energy();
+        }
```

Comment:
I find that this does not help for stopping the very lower energy e+/e- after along step actions, range/msc-limiter/boundary/msc-scatter/eloss (even though range is treated separately before along_step in celeritas).  So, let's apply `eloss_calc_limit` before `eloss.calc_eloss` (i.e., calculate the eloss above this cut, otherwise, eloss = particle energy as before). The rest looks okay to me and can be merged after @amandalund's review and approval.  As a follow up, let's review the whole e+/e- workflow and discuss how to properly stop very low energy e-/e+ below a certain energy threshold (a.k.a tracking cut) if necessary.  I can prepare a couple of slides for a follow up discussion at the next standup meeting.

Reply:
I'm confused. Applying before `eloss.calc_eloss` is what already is in the code, and it allows tracks that start *above* the cut to end up *below* the cut after slowing down, leading to low-energy tracks. Isn't this change (2507103c8b0c881613b4437cb2f22c33c9b9f9bd) what we discussed?

Reply:
Yes, my suggestion is reverting the change (i.e., back to the original).  I thought that the source of the very low energy particle is from the case when the calculated eloss is very close to the particle energy (above the eloss_calc_limit cut) and the particle with a small remaining energy is queued to the next step, but it seems that there is something else - so, I propose not to impose the additional relaxation on the eloss condition until a real source of missing logic in the work flow is understood.

Reply:
OK, maybe the confusion is about when and where propagation versus energy loss happens?

>  I thought that the source of the very low energy particle is from the case when the calculated eloss is very close to the particle energy (above the eloss_calc_limit cut) and the particle with a small remaining energy is queued to the next step

That is what I believe is happening. In the current implementation, a particle that ends a step with very low energy will undergo propagation (on the next step) before it hits the `eloss_limit` criteria.

Reply:
I believe that the current order of along step actions is right (i.e, the energy loss comes after the propagation and msc), but the status of track with the very small energy particle is not handled properly in the down stream (i.e., not either killed or stopped after depositing all its energy, so it still returns to the next along step even with this proposed change), which need to be understood.  Sorry for the confusion.

Reply:
Stopped positrons can reach the next step, but they will exit the along-step kernel early and not be propagated/undergo eloss etc.

I do see a very significant reduction in the number of very low energy (but not stopped) charged particles with this change (e.g. >99% fewer particles with energies below 1 eV).

Reply:
The stopped particle with `zero step length` exits at the beginning, which is good as it may be equivalently to the `zero energy` (stopped) particle (are these two are always exchangeable in celeritas?).

Okay, good to know that the cut rejects more than 99% of particle below 1eV.  Do you also know, what is the distribution of delta(energy-eloss) for primary particle energy below 1keV (eloss_calc_limit) and above 1keV.  The real question is why the new cut does not reject all particles below 1eV?  Who are those remaining low energy particles and where are they from?

Reply:
The cut only applies to tracks when:
- undergoing energy loss (i.e. not after a discrete interaction, nor to gammas)
- aren't in the process of hitting a boundary (so that we don't stop particles on boundaries)

So we're more likely to have small-energy particles starting their first step from the boundary, but they're likely to die after the next propagation.

Reply:
> The stopped particle with zero step length exits at the beginning, which is good as it may be equivalently to the zero energy (stopped) particle (are these two are always exchangeable in celeritas?).

Not sure about always, but yes at the beginning of along-step they are equivalent.

Reply:
Looking at what Geant4 does: to be equivalent we would also need to apply the tracking cut at the boundaries, and after discrete interactions. I think the latter is reasonable but we shouldn't do the former (to avoid biasing gamma production to happen on geometry edges, which also creates problems with particle tracking in ORANGE). I think we end up doing a half-step less work than Geant4 if we apply the cut as this PR implements.

Reply:
> Looking at what Geant4 does: to be equivalent we would also need to apply the tracking cut at the boundaries, and after discrete interactions

Where are you seeing this? It looks like if it's an energy loss processes and the particle's energy is below this cutoff the discrete interaction is skipped, but I don't see where the cut is applied after an interaction.

Also, I had originally thought Geant4 only applied this cut at the beginning of along-step (based on the pre-step energy), but actually it looks like they apply it _again_ after slowing down, so I guess this is what we should have been doing all along.

Reply:
Well damn that would explain it. I meant to go back and look at that code yesterday evening... I'm going to do another search for `G4EmParameters::LowestElectronEnergy` and see if that pops up anywhere.

Reply:
Looks like it's used in the G4EmCalculator cuts but nowhere else. It looks like the change I've proposed is equivalent to the "post-slowing-down" application of the cut.

Reply:
I am a bit confused - the lowest energy cut is used before the actual eloss calculation [here](https://geant4.kek.jp/lxr/source/processes/electromagnetic/utils/src/G4VEnergyLossProcess.cc#L812) and [after](https://geant4.kek.jp/lxr/source/processes/electromagnetic/utils/src/G4VEnergyLossProcess.cc#L914), which may not be same or equivalent to the proposed change.